tests/AnalyzeProperties.hs

```diff
@@ -0,0 +1,575 @@
+  pure $ mkDecimal $ Decimal.Decimal places mantissa
+
+genInteger :: MonadGen m => NumSize -> m Integer
+genInteger size = Gen.integral $
+  Range.exponentialFrom
+    (round   (midpoint size))
+    (ceiling (inf size))
+    (floor   (sup size))
+
+genArithOp :: MonadGen m => m ArithOp
+genArithOp = Gen.element [Add, Sub, Mul, Div] -- Pow, Log
+
+genUnaryArithOp :: MonadGen m => m UnaryArithOp
+genUnaryArithOp = Gen.element [Negate, Abs] -- Sqrt, Ln, Exp, Signum
+
+genRoundingLikeOp :: MonadGen m => m RoundingLikeOp
+genRoundingLikeOp = Gen.element [Round, Ceiling, Floor]
+
+genComparisonOp :: MonadGen m => m ComparisonOp
+genComparisonOp = Gen.element [Gt, Lt, Gte, Lte, Eq, Neq]
+
+type ECore = Existential (Core Analyze.Term)
+
+-- TODO: we might want to generalize this to a union of open intervals
+
+-- 'NumSize' bounds the values that a generated expression can evaluate to. We
+-- use this to avoid generating large numbers in computationally expensive
+-- spots.
+type NumSize = Interval Double
+
+data SizedType where
+  SizedInt     :: NumSize -> SizedType
+  SizedDecimal :: NumSize -> SizedType
+  SizedString  :: Int     -> SizedType
+  SizedTime    ::            SizedType
+  SizedBool    ::            SizedType
+  SizedKeySet  ::            SizedType
+  -- TODO: objects
+
+-- sizedLog :: NumSize -> (NumSize, NumSize)
+-- sizedLog size =
+--   let newExp  = logBase 10 size
+--       newBase = size ** (1 / newExp)
+--   in (newExp, newBase)
+
+arithSize :: ArithOp -> NumSize -> (NumSize, NumSize)
+arithSize op size = case op of
+  Add -> (size, size)
+  Sub -> (size, size)
+  Mul -> (sqrt size, sqrt size)
+  Div -> (size, size)
+  Pow -> undefined
+  Log -> undefined
+
+unaryArithSize :: UnaryArithOp -> NumSize -> NumSize
+unaryArithSize op size = case op of
+  Negate -> - size
+  Sqrt   -> size ** 2
+  Abs    -> size -- ?
+  Ln     -> undefined
+  Exp    -> undefined
+  Signum -> undefined
+
+mkInt :: MonadGen m => Core Analyze.Term Integer -> m ETerm
+mkInt = pure . ESimple TInt . Inj
+
+mkDec :: MonadGen m => Numerical Analyze.Term Decimal -> m ETerm
+mkDec = pure . ESimple TDecimal . Inj . Numerical
+
+class Extract a where
+  extract :: ETerm -> Analyze.Term a
+
+instance Extract Integer where
+  extract = \case
+    ESimple TInt x -> x
+    other -> error (show other)
+
+instance Extract Decimal where
+  extract = \case
+    ESimple TDecimal x -> x
+    other -> error (show other)
+
+instance Extract String where
+  extract = \case
+    ESimple TStr x -> x
+    other -> error (show other)
+
+instance Extract Bool where
+  extract = \case
+    ESimple TBool x -> x
+    other -> error (show other)
+
+instance Extract Time where
+  extract = \case
+    ESimple TTime x -> x
+    other -> error (show other)
+
+-- TODO: Var, objects
+-- TODO: we might want to reweight these by using `Gen.frequency`.
+genCore :: MonadGen m => SizedType -> m ETerm
+genCore (SizedInt size) = Gen.recursive Gen.choice [
```

Comment:
It might also be possible generate these in a generic way, a la https://github.com/Lysxia/generic-random, but (1) that package doesn't exist for hedgehog, and (2) we don't have generic instances for these data types.

Reply:
This is one of the two large pieces of boilerplate in this file.